### PR TITLE
fix(server): Exclude eval from pipeline squashing

### PR DESCRIPTION
**Pull request by dranikpg**
* exclude eval from pipeline squashing
* add some debug prints to multi-command-squasher

### CODE REVIEW

File: src/server/main_service.cc [hunk 1]
```diff
@@ -1101,10 +1101,16 @@ void Service::DispatchManyCommands(absl::Span<CmdArgList> args_list,
 
     // MULTI...EXEC commands need to be collected into a single context, so squashing is not
     // possible
-    const bool is_multi =
+    bool is_multi =
         dfly_cntx->conn_state.exec_info.IsCollecting() || CO::IsTransKind(ArgS(args, 0));
 
-    if (!is_multi && cid != nullptr) {
+    // Generally, executing any multi-transactions (including eval) is not possible because they
+    // migh request a stricter multi mode than non-atomic which is used for squashing.
```
Comment (chakaz) [inline src/server/main_service.cc:1108]: migh->might

File: src/server/main_service.cc [hunk 2]
```diff
@@ -1101,10 +1101,16 @@ void Service::DispatchManyCommands(absl::Span<CmdArgList> args_list,
 
     // MULTI...EXEC commands need to be collected into a single context, so squashing is not
     // possible
-    const bool is_multi =
+    bool is_multi =
```
Comment (chakaz) [inline src/server/main_service.cc:1104]: Why not have `is_multi` be `const`? Am I missing a mutation somewhere?

File: src/server/main_service.cc [hunk 3]
```diff
@@ -1101,10 +1101,16 @@ void Service::DispatchManyCommands(absl::Span<CmdArgList> args_list,
...
         dfly_cntx->conn_state.exec_info.IsCollecting() || CO::IsTransKind(ArgS(args, 0));
 
-    if (!is_multi && cid != nullptr) {
+    // Generally, executing any multi-transactions (including eval) is not possible because they
+    // migh request a stricter multi mode than non-atomic which is used for squashing.
+    // TODO: By allowing promotion we can potentially execute multiple eval in parallel, which is
+    // very powerful paired with shardlocal eval
```
Comment (chakaz) [inline src/server/main_service.cc:1110]: I don't follow this comment. Promotion as in - allow squashing?
In the past you said that squashing doesn't make sense for single shard lua, I think

File: src/server/main_service.cc [hunk 4]
```diff
@@ -1594,13 +1600,15 @@ void Service::EvalInternal(CmdArgList args, const EvalArgs& eval_args, Interpret
...
 
-  optional<ShardId> sid;
-
   // TODO: to determine whether the script is RO by scanning all "redis.p?call" calls
   // and checking whether all invocations consist of RO commands.
   // we can do it once during script insertion into script mgr.
   auto& sinfo = cntx->conn_state.script_info;
   sinfo = make_unique<ConnectionState::ScriptInfo>();
+
+  absl::Cleanup sinfo_clean = [&sinfo]() { sinfo.reset(); };
```
Comment (chakaz) [inline src/server/main_service.cc:1609]: Nit: we could add `sinfo.reset()` to the below `clean` object

File: src/server/main_service.cc (hunk 2 above)
Comment (dranikpg) [reply src/server/main_service.cc:1104]: I tried to fit it into a single line 😄

File: src/server/main_service.cc (hunk 3 above)
Comment (dranikpg) [reply src/server/main_service.cc]: Fixed the comment, made it more detailed

**Review by chakaz — APPROVED** [review]

**Pull request merged**